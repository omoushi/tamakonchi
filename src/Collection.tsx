import React, {FC} from "react";
import {View, Text} from "react-native";

export const Collection: FC<{title: 'Detail'}> = props => (
  <View>
    <Text>{props.title}</Text>
  </View>
);

export default Collection