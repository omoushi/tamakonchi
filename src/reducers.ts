import { combineReducers } from "redux";
import { main } from "./main/reducer";
import { MainState } from "./main/interface";
import { TimerState } from "./timer/interface";
import { timer } from "./timer/reducer";

export interface RootState {
  main: MainState;
  timer: TimerState;
}

export default combineReducers<RootState>({
  main,
  timer,
});
