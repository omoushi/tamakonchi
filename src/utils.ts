export function randomIn<E>(array: E[]): E {
  return array[Math.floor(Math.random() * array.length)];
}