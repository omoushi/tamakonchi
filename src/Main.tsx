import React, { FC, ReactElement } from "react";
import { Button, StyleSheet, Text, View } from "react-native";
import { MainEvents, MainState } from "./main/interface";
import { Action, Dispatch } from "redux";
import { breakUp, loseJob, makeWorry, neglect } from "./main/actions";
import { connect } from "react-redux";
import { RootState } from "./reducers";
import { Knife } from "./components/Knife";
import { TimerState } from "./timer/interface";
import { NavigationTransitionProps } from "react-navigation";
import { Situation } from "./main/reducer";

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#F5FCFF',
  },
  header: {
    flexDirection: 'row',
    justifyContent: 'flex-start',
  },
  contents: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  }
});

type MainProps = MainEvents & MainState & NavigationTransitionProps & TimerState;

export const Main: FC<MainProps> = (props: MainProps): ReactElement => {
  const isSicked = props.situation !== Situation.EMPTY;
  return (
    <View style={styles.container}>
      <View style={styles.header}>
        <Button
          title={'おもひで'}
          onPress={() => {
            props.navigation.navigate('Collection');
          }}
        />
      </View>
      <View style={styles.contents}>
        <Text>{props.stage}</Text>
        <Text>{props.situation}</Text>
        {isSicked ? <Text>病んでる</Text> : null}
        <Button title={'別れる'} onPress={props.breakUp}/>
        <Button title={'放置する'} onPress={props.neglect}/>
        <Button title={'仕事がなくなる'} onPress={props.loseJob}/>
        <Button title={'心配にさせる'} onPress={props.makeWorry}/>
        {props.tools.map((tool): ReactElement => <Knife key={tool.id} tool={tool} />)}
      </View>
    </View>
  );
};

export const mapStateToProps = (state: RootState): MainState & TimerState => ({...state.main, ...state.timer});
export const mapDispatchToProps = (dispatch: Dispatch): MainEvents => ({
  breakUp: (): Action => dispatch(breakUp()),
  neglect: (): Action => dispatch(neglect()),
  loseJob: (): Action => dispatch(loseJob()),
  makeWorry: (): Action => dispatch(makeWorry())
});

export default connect(mapStateToProps, mapDispatchToProps)(Main);