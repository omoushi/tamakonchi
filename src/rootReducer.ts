import { combineReducers } from "redux";

import { MainState } from "./types/main";
import { TimerState } from "./types/timer";
import { main } from "./reducers/main";
import { timer } from "./reducers/timer";

export interface RootState {
  main: MainState;
  timer: TimerState;
}

export default combineReducers<RootState>({
  main,
  timer,
});
