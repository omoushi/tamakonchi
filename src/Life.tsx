import React, {FC} from "react";
import {Button, Text, View} from "react-native";
import {Action, Dispatch, Reducer} from "redux";
import {connect} from "react-redux";
import {randomIn} from "./utils";

type Props = State & Events
type State = {
  status: 'たまご' | 'いきてる' | 'しんでる'
  health: '健康' | '不健康' | '不明'
}
type Events = {
  takeCare(): void
  notTakeCare(): void
}

const initialState: State = {
  status: 'たまご',
  health: '健康'
};

export enum Statuses {
  STATUS, HEALTH
}

export const reducer: Reducer = (state: State = initialState, action: Action): State => {
  switch (action.type) {
    case 'TAKE_CARE':
      switch (state.status) {
        case 'たまご':
          return {...state, status: 'いきてる'};
        case 'いきてる':
        case 'しんでる':
        default:
          return {...state}
      }
    case 'NOT_TAKE_CARE':
      switch (randomIn([Statuses.STATUS, Statuses.HEALTH])) {
        case Statuses.STATUS:
          switch (state.status) {
            case 'たまご':
            case 'いきてる':
              return {...state, status: 'しんでる', health: '不明'};
            case "しんでる":
            default:
              return {...state}
          }
        case Statuses.HEALTH:
          if (state.health === '健康') {
            return {...state, health: '不健康'};
          } else {
            return {...state}
          }
        default:
          return {...state}
      }
    default:
      return {...state}
  }
};

const Life: FC<Props> = props => (
  <View>
    <Text>{props.status}</Text>
    <Text>{props.health}</Text>
    <Button title={'世話をする'} onPress={props.takeCare}/>
    <Button title={'世話をしない'} onPress={props.notTakeCare}/>
  </View>
);
const mapStateToProps = (state: State) => state;
const mapDispatchToProps = (dispatch: Dispatch): Events => ({
  takeCare: () => dispatch({type: 'TAKE_CARE'}),
  notTakeCare: () => dispatch({type: 'NOT_TAKE_CARE'})
});
export default connect(mapStateToProps, mapDispatchToProps)(Life)