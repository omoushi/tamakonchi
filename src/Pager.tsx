import React, { ReactNode } from "react";
import { StyleSheet, View, Button } from "react-native";
import Main from "./Main";
import { connect } from "react-redux";
import Collection from "./Collection";
import { createStackNavigator, createAppContainer } from 'react-navigation';


const RootStack = createStackNavigator(
  {
    Main: {
      screen: Main,
    },
    Collection: {
      screen: Collection,
    },
  },
  {
    initialRouteName: 'Main',
  },
);

const Navigation = createAppContainer(RootStack);
class Pager extends React.Component {
  private styles = StyleSheet.create({
    container: {
      flex: 1,
      flexDirection: 'column',
      justifyContent: 'center',
      backgroundColor: '#F5FCFF',
    },
  });

  public componentDidMount(): void {
    console.log('Pager');
  }

  public componentWillUnmount(): void {
    console.log('Pager');
  }
  public render(): ReactNode {
    return (
      <View style={this.styles.container}>
        <Navigation/>
      </View>
    );
  }
}

export default connect()(Pager);