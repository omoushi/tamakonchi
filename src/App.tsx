/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 */

import React, { ReactNode } from "react";
import {createStore} from "redux";
import {Provider} from "react-redux";
import reducer from "./reducers";
import Pager from "./Pager";

const store = createStore(reducer);

const RootStack = createMaterialTopTabNavigator(
  {
    Main: {
      screen: Main,
    },
    Collection: {
      screen: Collection,
    },
  },
  {
    initialRouteName: 'Main',
    tabBarOptions: {
      tabStyle: {
        marginTop: 20,
      },
    },
  },
);

const Navigation = createAppContainer(RootStack);

export default class App extends React.Component {
  public render(): ReactNode {
    return (
      <Provider store={store}>
        <Pager/>
      </Provider>
    );
  }
}
