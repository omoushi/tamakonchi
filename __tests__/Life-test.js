import {reducer, Statuses} from "../src/Life";
import React from "react";
import * as utils from '../src/utils';

const statePatterns = [
  undefined,
  {status: 'たまご', health: '健康'},
  {status: 'たまご', health: '不健康'},
  {status: 'いきてる', health: '健康'},
  {status: 'いきてる', health: '不健康'},
  {status: 'しんでる', health: '不明'}
];

describe('takeCare', () => {
  const takeCareAction = {type: 'TAKE_CARE'};

  statePatterns.forEach(state => {
    it(`state: ${JSON.stringify(state)}`, () => {
      expect(reducer(state, takeCareAction)).toMatchSnapshot()
    });
  });
});

describe('notTakeCare', () => {
  const notTakeCareAction = {type: 'NOT_TAKE_CARE'};
  jest.mock('../src/utils');

  describe('状態が変わる場合', () => {
    utils.randomIn = jest.fn(() => Statuses.STATUS);

    statePatterns.forEach(state => {
      it(`state: ${JSON.stringify(state)}`, () => {
        expect(reducer(state, notTakeCareAction)).toMatchSnapshot()
      });
    });
  });

  describe('健康が変わる場合', () => {
    utils.randomIn = jest.fn(() => Statuses.HEALTH);

    statePatterns.forEach(state => {
      it(`state: ${JSON.stringify(state)}`, () => {
        expect(reducer(state, notTakeCareAction)).toMatchSnapshot()
      });
    });
  });
});