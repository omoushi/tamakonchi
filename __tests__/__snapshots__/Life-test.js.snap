// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`notTakeCare 健康が変わる場合 state: {"status":"いきてる","health":"不健康"} 1`] = `
Object {
  "health": "不健康",
  "status": "いきてる",
}
`;

exports[`notTakeCare 健康が変わる場合 state: {"status":"いきてる","health":"健康"} 1`] = `
Object {
  "health": "不健康",
  "status": "いきてる",
}
`;

exports[`notTakeCare 健康が変わる場合 state: {"status":"しんでる","health":"不明"} 1`] = `
Object {
  "health": "不明",
  "status": "しんでる",
}
`;

exports[`notTakeCare 健康が変わる場合 state: {"status":"たまご","health":"不健康"} 1`] = `
Object {
  "health": "不健康",
  "status": "たまご",
}
`;

exports[`notTakeCare 健康が変わる場合 state: {"status":"たまご","health":"健康"} 1`] = `
Object {
  "health": "不健康",
  "status": "たまご",
}
`;

exports[`notTakeCare 健康が変わる場合 state: undefined 1`] = `
Object {
  "health": "不健康",
  "status": "たまご",
}
`;

exports[`notTakeCare 状態が変わる場合 state: {"status":"いきてる","health":"不健康"} 1`] = `
Object {
  "health": "不健康",
  "status": "いきてる",
}
`;

exports[`notTakeCare 状態が変わる場合 state: {"status":"いきてる","health":"健康"} 1`] = `
Object {
  "health": "不健康",
  "status": "いきてる",
}
`;

exports[`notTakeCare 状態が変わる場合 state: {"status":"しんでる","health":"不明"} 1`] = `
Object {
  "health": "不明",
  "status": "しんでる",
}
`;

exports[`notTakeCare 状態が変わる場合 state: {"status":"たまご","health":"不健康"} 1`] = `
Object {
  "health": "不健康",
  "status": "たまご",
}
`;

exports[`notTakeCare 状態が変わる場合 state: {"status":"たまご","health":"健康"} 1`] = `
Object {
  "health": "不健康",
  "status": "たまご",
}
`;

exports[`notTakeCare 状態が変わる場合 state: undefined 1`] = `
Object {
  "health": "不健康",
  "status": "たまご",
}
`;

exports[`takeCare state: {"status":"いきてる","health":"不健康"} 1`] = `
Object {
  "health": "不健康",
  "status": "いきてる",
}
`;

exports[`takeCare state: {"status":"いきてる","health":"健康"} 1`] = `
Object {
  "health": "健康",
  "status": "いきてる",
}
`;

exports[`takeCare state: {"status":"しんでる","health":"不明"} 1`] = `
Object {
  "health": "不明",
  "status": "しんでる",
}
`;

exports[`takeCare state: {"status":"たまご","health":"不健康"} 1`] = `
Object {
  "health": "不健康",
  "status": "いきてる",
}
`;

exports[`takeCare state: {"status":"たまご","health":"健康"} 1`] = `
Object {
  "health": "健康",
  "status": "いきてる",
}
`;

exports[`takeCare state: undefined 1`] = `
Object {
  "health": "健康",
  "status": "いきてる",
}
`;
